core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java

```diff
@@ -17,30 +17,36 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(AmbariHostsJoinStatusCheckerTask.class);
 
     @Override
-    public boolean checkStatus(AmbariHostsCheckerContext hosts) {
+    public boolean checkStatus(AmbariHostsCheckerContext ambariHostsCheckerContext) {
         try {
-            AmbariClient ambariClient = hosts.getAmbariClient();
-            Map<String, String> hostNames = ambariClient.getHostStatuses();
-            for (HostMetadata hostMetadata : hosts.getHostsInCluster()) {
-                boolean contains = false;
-                for (Entry<String, String> hostName : hostNames.entrySet()) {
-                    if (hostName.getKey().equals(hostMetadata.getHostName()) && !"UNKNOWN".equals(hostName.getValue())) {
-                        contains = true;
-                        break;
-                    }
-                }
-                if (!contains) {
-                    LOGGER.info("The host {} currently not part of the cluster, waiting for join", hostMetadata.getHostName());
-                    return false;
-                }
-            }
-        } catch (Exception ignored) {
-            LOGGER.info("Did not join all hosts yet, polling");
+            return areAllHostsJoined(ambariHostsCheckerContext);
+        } catch (Exception e) {
+            LOGGER.info("Not all hosts joined yet. Continuing polling.", e);
             return false;
         }
+    }
+
+    private boolean areAllHostsJoined(AmbariHostsCheckerContext ambariHostsCheckerContext) {
+        AmbariClient ambariClient = ambariHostsCheckerContext.getAmbariClient();
+        Map<String, String> hostNamesToStatuses = ambariClient.getHostStatuses();
+        for (HostMetadata host : ambariHostsCheckerContext.getHostsInCluster()) {
+            if (!isHostJoined(host, hostNamesToStatuses)) {
+                LOGGER.info("Host {} is currently not part of the cluster, waiting for it to join.", host.getHostName());
+                return false;
```

Comment:
here is also worth to refactor to stream with filter, in this case we can log all of the hosts which isn't joined, not just the first

Reply:
You are right, this one worth touching. (Also just extracted.)

Reply:
I am also writing tests for it.